### PR TITLE
Gate cloning undo

### PR DESCRIPTION
**Pull request by openroadie**
Add support for undo with gate cloning. Updated some ok files

### CODE REVIEW

**Comment by openroadie** [comment]
ORFS results with this code will be available here:

https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1196

File: src/rsz/include/rsz/Resizer.hh [hunk 1]
```diff
@@ -639,7 +640,8 @@ protected:
   InstanceSeq inserted_buffers_;
   InstanceSet inserted_buffer_set_;
   Map<Instance *, std::tuple<LibertyPort *, LibertyPort *>> swapped_pins_;
-  Map<Instance *, Instance *> cloned_gates_;
+  std::stack<std::tuple<Instance *, Instance *>> cloned_gates_;
```
Comment (QuantamHD) [inline src/rsz/include/rsz/Resizer.hh:643]: Could give a using name to std::tuple<Instance *, Instance *>. I feel that it would help readability here.

**Comment by QuantamHD** [comment]
If possible could you add some C++ unit tests to this PR. It feels like it would help tease out issues in the undo function.

**Comment by openroadie** [comment]
> If possible could you add some C++ unit tests to this PR? It feels like it would help tease out issues in the undo function.

I will take a look. Given the fact that the problems with undo were mostly related to the combination of buffer insertion/deleteion/resize/cloning.... I am not too sure about two things:

- Whether writing a unit test as opposed to the way we test right now is the most efficient way to ensure quality.
- Writing a unit test is feasible.

File: src/rsz/src/Resizer.cc [hunk 2]
```diff
@@ -2652,7 +2652,10 @@ Resizer::journalBegin()
   debugPrint(logger_, RSZ, "journal", 1, "journal begin");
   resized_inst_map_.clear();
   inserted_buffers_.clear();
-  cloned_gates_.clear();
+  inserted_buffer_set_.clear();
+  while(!cloned_gates_.empty())
+    cloned_gates_.pop();
```
Comment (maliberty) [inline src/rsz/src/Resizer.cc:2657]: cloned_gates_ = {};

File: src/rsz/src/Resizer.cc [hunk 3]
```diff
@@ -2662,7 +2665,10 @@ Resizer::journalEnd()
   debugPrint(logger_, RSZ, "journal", 1, "journal end");
   resized_inst_map_.clear();
   inserted_buffers_.clear();
-  cloned_gates_.clear();  
+  inserted_buffer_set_.clear();
+  while(!cloned_gates_.empty())
+    cloned_gates_.pop();
```
Comment (maliberty) [inline src/rsz/src/Resizer.cc:2670]: same

**Review by maliberty — APPROVED** [review]

**Comment by maliberty** [comment]
Conflicts to resolve

**Comment by maliberty** [comment]
CI failures...

**Pull request merged**